activerecord/lib/active_record.rb

```diff
@@ -381,6 +381,50 @@ def self.global_executor_concurrency # :nodoc:
   singleton_class.attr_accessor :dump_schemas
   self.dump_schemas = :schema_search_path
 
+  singleton_class.attr_accessor :db_config_handlers # :nodoc:
+  self.db_config_handlers = [] # :nodoc:
+
+  # Allows an application to register a custom handler for database configuration
+  # objects. This is useful for creating a custom handler that responds to
+  # methods your application needs but Active Record doesn't implement. For
+  # example if you are using Vitess, you may want your Vitess configurations
+  # to respond to `sharded?`. To implement this define the following in an
+  # initializer:
+  #
+  #   ActiveRecord.register_db_config_handler do |env_name, name, url, config|
+  #     next unless config.key?(:vitess)
+  #     VitessConfig.new(env_name, name, config)
+  #   end
+  #
+  # Note: applications must handle the condition in which custom config should be
+  # created in your handler registration otherwise all objects will use the custom
+  # handler.
+  #
+  # Then define your +VitessConfig+ to respond to the methods your application
+  # needs. It is recommended that you inherit from one of the existing
+  # database config classes to avoid having to reimplement all methods. Custom
+  # config handlers should only implement methods Active Record does not.
+  #
+  #   class VitessConfig < ActiveRecord::DatabaseConfigurations::UrlConfig
+  #     def sharded?
+  #       configuration_hash.fetch("sharded", false)
+  #     end
+  #   end
+  #
+  # For configs that have a +:vitess+ key, a +VitessConfig+ object will be
+  # created instead of a +UrlConfig+.
+  def self.register_db_config_handler(&block)
+    db_config_handlers << block
+  end
+
+  register_db_config_handler do |env_name, name, url, config|
+    if url
+      DatabaseConfigurations::UrlConfig.new(env_name, name, url, config)
+    else
```

Comment:
Would it make sense to e.g. `elsif config.key?(:adapter) || config.key?("adapter")` here, and then have `build_db_config_from_hash` raise if it falls out of the loop without any registered handlers matching?

Reply:
We can't do this currently because at the end we merge in the database_url, after the hash configs are created. I'm separately working on moving that earlier so that we don't need to do that. But at the moment we'd end up raising for configs that will eventually be perfectly fine. See https://github.com/rails/rails/blob/main/activerecord/lib/active_record/database_configurations.rb#L164